### PR TITLE
Introduce external account provider into `pallet_evm`

### PR DESCRIPTION
**Pull request by dmitrylavrenov**
To simplify #60.

The frontier tests have been passed successfully.

### CODE REVIEW

File: frame/evm/src/lib.rs [hunk 1]
```diff
@@ -913,3 +916,45 @@ impl<T> OnCreate<T> for Tuple {
...
+/// indexes (the number of previous transactions associated with an account).
+pub trait AccountProvider {
+	/// The user account identifier type.
+	type AccountId;
+	/// Account index (aka nonce) type.
+	type Index: AtLeast32Bit;
+
+	/// Creates a new account.
+	fn create_account(who: &Self::AccountId);
+	/// Removed an account.
```
Comment (MOZGIII) [inline frame/evm/src/lib.rs:932]:
```suggestion
	/// Removes an account.
```

File: frame/evm/src/lib.rs [hunk 2]
```diff
@@ -913,3 +916,45 @@ impl<T> OnCreate<T> for Tuple {
...
+	fn remove_account(who: &Self::AccountId);
+	/// Return current account nonce value.
+	fn account_nonce(who: &Self::AccountId) -> Self::Index;
+	/// Increment account nonce value.
+	fn inc_account_nonce(who: &Self::AccountId);
+}
+
+pub struct NativeSystemAccountProvider<T>(sp_std::marker::PhantomData<T>);
```
Comment (MOZGIII) [inline frame/evm/src/lib.rs:940]: Missing the docs here.

File: frame/evm/src/lib.rs (hunk 1 above)
Comment (dmitrylavrenov) [reply frame/evm/src/lib.rs]: https://github.com/humanode-network/frontier/pull/61/commits/8f52f7ee8e220647ee5cde9ef4c294677c0d5d1f

File: frame/evm/src/lib.rs (hunk 2 above)
Comment (dmitrylavrenov) [reply frame/evm/src/lib.rs:951]: Added

**Comment by dmitrylavrenov** [comment]
> The code structure looks fine!
> 
> A few notes:
> 
> * We might want to push the trait declaration to a separate crate to keep cleaner abstraction isolation.
> * The trait documentation should sufficiently explain what the trait is intended for, and each method's docs explain what it does, when it can be called and why, etc. This is tricky because we can't be too specific as to not refer to a particular implementation, which can possibly update out of sync from the documentation we currently discuss. So we need to be quite abstract in regards to mentioning the code parts (i.e. function names and etc) and explain things in terms of general logic instead.
> 
> Let's discuss if in doubt. Otherwise, this very looks good!
> 
> Oh, and one more thing: we need to adjust the rest of the crates too: precompiles, etc... Or is it just all we need?

The trait related docs have been improved. Your suggestion would be great as well!
Moved to separate mod as I was thinking that it's close part of the `pallet_evm` itself as `runner` logic for example. Would be happy to understand if it should be declared in separate crate.

**Review by MOZGIII — APPROVED** [review]

**Pull request merged**